### ISSUE TITLE
Support using a custom WebSocketServer class

### ISSUE THREAD

**Issue by cjerdonek** [aaugustin/websockets#230]
This issue is about letting [`serve()`](https://github.com/aaugustin/websockets/blob/4703884a5c82452c25ed39224160913162cf448e/websockets/server.py#L409) use a custom `WebSocketServer` class.

When I filed issue #216, I was interested in giving every `WebSocketServerProtocol` object access to a value computed prior to calling `serve()`. An earlier idea was to subclass `WebSocketServerProtocol` and add a custom instance attribute. (Another idea was to add a class attribute to `WebSocketServerProtocol`, but that seems hacky to me because the value is computed dynamically.)

However, since `serve()` passes the `WebSocketServer` object to every `WebSocketServerProtocol` object it creates, it seems like the better approach is to add the custom attribute to the `WebSocketServer` class, and then have the protocol objects access the attribute through their `ws_server` attribute.

The easiest / most natural thing to do would be to add to `serve()` either--

1. a `ws_server_klass` argument (or better yet perhaps `make_ws_server` -- see issue #216) that defaults to `WebSocketServer`, or
2. a `ws_server` argument that defaults to `WebSocketServer(loop)`.

**Comment by aaugustin** [comment]
I'll think about this.

I have a backlog of PRs of yours waiting for my input, and I'd like to finish #178 too, so please bear with me while I work through all this... ;-)

**Comment by cjerdonek** [comment]
No prob! And thanks for your time with all of these!

**Comment by cjerdonek** [comment]
To match the `ws_server` attribute and argument name and so as not to confuse with the `server` argument and attribute name (which is an `asyncio.Server` object), I would suggest either `ws_server` if passing an instance, or `ws_server_factory` if passing a callable (to parallel `protocol_factory` suggested [here](https://github.com/aaugustin/websockets/issues/216#issuecomment-317281879)).

**Comment by cjerdonek** [comment]
With issue #216 being implemented as `create_protocol`, `create_ws_server` seems to me like the most natural candidate for this issue. Like `create_protocol`, the argument for this parameter could be either a class or callable, and the class / callable should accept the same arguments as `WebSocketServer`'s constructor. The return value would get set to the `ws_server` attribute of each `WebSocketServerProtocol` instance created.

**Comment by aaugustin** [comment]
I tend to prefer the functional approach discussed in #271, but I haven't looked at this deeply yet.

**Comment by cjerdonek** [comment]
This issue is different from #271 though. The issue here is that `serve()` hardcodes `WebSocketServer`. 

The connection to #271 I mentioned is that a custom server class could be used to give a handler function access to other values, since the server object is already being passed to the protocol object when it's first constructed anyways (and in particular before the handler function is called, which I believe makes this a more general solution). But the reverse isn't true: using a partial won't let you patch in a custom server class.

**Comment by aaugustin** [comment]
Okay! I should really take some time to get to the bottom of this... likely after 4.0.

**Comment by aaugustin** [comment]
Sorry for the delay :-(

----

First, let me rephrase my understanding of this issue.

For example, your protocol needs access to a `sheep = make_herd_of_sheep()` variable. Every protocol instance uses the same `sheep`. Protocol instances need a way to access this shared variable.

It should be reasonably encapsulated: defining it at the module level or as a class attribute would be inelegant (and possibly make testing more difficult).

After #216, the solution looks like:

```python
create_protocol = functools.partial(
    WebSocketServerProtocol,
    sheep=make_herd_of_sheep()
)
start_server = websockets.serve(
   ws_handler, host, port, create_protocol=create_protocol)
```

This is what I called the "functional approach" and I like it.

----

Then you noticed that there's already some shared state between all protocol instances running in the same server: the `WebSocketServer` instance available in `self.ws_server`.

Until now, `WebSocketServer` is only documented as "the thing that `serve()` returns, and which provides `close()`, `wait_close()` and `sockets`".

It's an opaque and non-customizable object, because we haven't had a need to allow customizing it.

Allowing users to subclass `WebSocketServer` and use that subclass would be an "object-oriented approach".

----

I have a preference for the functional approach and reservations about the object-oriented approach:

- the current design remains virtually identical to asyncio's [create_server](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.AbstractEventLoop.create_server) and [Server](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.Server) APIs. That's good because it makes knowledge easier to transfer between `asyncio` and `websockets`, which improves user experience.

- Subclassing as an API always leads to hard-to-maintain code (ever heard people complain about Django's generic class based-views? there's a whole website dedicated to helping people figure out how to subclass them: https://ccbv.co.uk). Unfortunately asyncio Protocols require subclassing; this decision bleeds on websockets and we end up having to expose that ability, reluctantly. I don't want to encourage more subclassing when there's an alternate solution, and here we have one.

- Documenting `WebSocketServerProtocol.ws_server` locks that implementation detail for the future. I'm not very worried but I'd rather avoid it if we can.

- I'm afraid users may add all their state-managing logic in a `WebSocketServer` subclass and I suspect they may regret it. If they start doing this and define a `register` or `unregister` method, oops... everything breaks. I'd rather keep `WebSocketServer` opaque and non-customizable. And it's certainly better to put the state management in a standalone piece of code that can be tested in isolation.

----

I understand your argument about mirroring `create_protocol` with `create_ws_server`, however my dislike for "subclassing as an API" is greater than my desire for consistency :-) and I haven't seen a sufficiently large advantage in customizing `WebSocketServer`.
